File: streamflow-backend/app/api/routes/movies.py
from typing import Any

from fastapi import APIRouter, Query

from app.schemas.movie import GenreListResponse, MovieDetails, MovieListResponse
from app.services import movie_service

router = APIRouter()


@router.get("/genres", response_model=GenreListResponse)
async def get_movie_genres() -> GenreListResponse:
    return await movie_service.get_genres()  # type: ignore[return-value]


@router.get("/trending", response_model=MovieListResponse)
async def get_trending_movies(
    page: int = Query(1, ge=1, le=1000, description="Page number"),
) -> MovieListResponse:
    return await movie_service.get_trending_movies(page=page)


@router.get("/popular", response_model=MovieListResponse)
async def get_popular_movies(
    page: int = Query(1, ge=1, le=1000, description="Page number"),
) -> MovieListResponse:
    return await movie_service.get_popular_movies(page=page)


@router.get("/top-rated", response_model=MovieListResponse)
async def get_top_rated_movies(
    page: int = Query(1, ge=1, le=1000, description="Page number"),
) -> MovieListResponse:
    return await movie_service.get_top_rated_movies(page=page)


@router.get("/search")
async def search_movies(
    query: str | None = Query(None, description="Search query"),
    page: int = Query(1, ge=1, le=1000, description="Page number"),
    genre: int | None = Query(None, description="Genre ID filter"),
    year: int | None = Query(None, ge=1900, le=2100, description="Release year filter"),
    rating: float | None = Query(None, ge=0, le=10, description="Minimum rating filter"),
    deep_search: bool = Query(False, description="Enable broad multi-search"),
) -> dict[str, Any]:
    return await movie_service.search_movies(
        query=query, 
        page=page, 
        genre=genre, 
        year=year, 
        rating=rating,
        deep_search=deep_search
    )


@router.get("/{tmdb_id}", response_model=MovieDetails)
async def get_movie_details(tmdb_id: int) -> MovieDetails:
    return await movie_service.get_movie_details(tmdb_id=tmdb_id)

File: streamflow-frontend/src/services/movies.js
import api from './api';
import { API_ENDPOINTS } from '../utils/constants';

export const getTrendingMovies = async (page = 1) => {
  const response = await api.get(API_ENDPOINTS.MOVIES.TRENDING, {
    params: { page },
  });
  return response.data;
};

export const getPopularMovies = async (page = 1) => {
  const response = await api.get(API_ENDPOINTS.MOVIES.POPULAR, {
    params: { page },
  });
  return response.data;
};

export const getTopRatedMovies = async (page = 1) => {
  const response = await api.get(API_ENDPOINTS.MOVIES.TOP_RATED, {
    params: { page },
  });
  return response.data;
};

export const getMovieDetails = async (tmdbId) => {
  const response = await api.get(API_ENDPOINTS.MOVIES.DETAILS(tmdbId));
  return response.data;
};

export const searchMovies = async (query, filters = {}, page = 1, deepSearch = false) => {
  const cleanFilters = Object.entries(filters).reduce((acc, [key, value]) => {
    if (value !== '' && value !== null && value !== undefined) {
      acc[key] = value;
    }
    return acc;
  }, {});

  const response = await api.get(API_ENDPOINTS.MOVIES.SEARCH, {
    params: { query, page, ...cleanFilters, deep_search: deepSearch },
  });
  return response.data;
};
